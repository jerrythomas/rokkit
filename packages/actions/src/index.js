export { fillable } from './fillable'
export { pannable } from './pannable'
