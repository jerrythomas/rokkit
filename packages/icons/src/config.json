{
  "package": {
    "namespace": "@rokkit",
    "homepage": "https://github.com/jerrythomas/rokkit"
  },
  "bundles": {
    "auth": { "color": true }
  }
}
