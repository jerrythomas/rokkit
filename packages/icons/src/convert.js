import {
	importDirectory,
	cleanupSVG,
	runSVGO,
	parseColors,
	isEmptyColor,
	exportJSONPackage
} from '@iconify/tools'
import fs from 'fs'

/**
 * Clean up and optimise SVG icon
 *
 * @param {string} svg SVG code
 * @param {boolean} color True if color should be preserved
 */
export function cleanAndOptimizeIcon(svg, color) {
	// Clean up icon code
	cleanupSVG(svg)

	if (!color) {
		parseColors(svg, {
			defaultColor: 'currentColor',
			callback: (_, colorStr, value) => {
				return !value || isEmptyColor(value) ? colorStr : 'currentColor'
			}
		})
	}

	// Optimise
	runSVGO(svg)
}

/**
 * Process icons
 *
 * @param {IconifyTools} iconSet Icon set
 * @param {boolean} color True if color should be preserved
 */
export function processIcons(iconSet, color) {
	iconSet.forEach((name, type) => {
		// There should not be any type other than icon
		// if (type !== 'icon') {
		// 	return
		// }

		const svg = iconSet.toSVG(name)
		// Since we are processing from within the list, there is no need to check for invalid icons
		// if (!svg) {
		// 	// Invalid icon
		// 	iconSet.remove(name)
		// 	return
		// }

		// Clean up and optimise icons
		try {
			cleanAndOptimizeIcon(svg, color)
		} catch (err) {
			// eslint-disable-next-line no-console
			console.error(`Error parsing ${name}: ${err.reason}`)
			iconSet.remove(name)
			return
		}

		// Update icon
		iconSet.fromSVG(name, svg)
	})
}

/**
 * Convert icons
 *
 * @param {string} folder Folder with icons
 * @param {string} prefix Prefix for icon set
 * @param {boolean} color True if color should be preserved
 */
export async function convert(folder, prefix, color = false) {
	// Import icons
	const iconSet = await importDirectory(folder, { prefix })

	// Validate, clean up, fix palette and optimise
	processIcons(iconSet, color)

	// Export
	const collection = JSON.stringify(iconSet.export(), null, 2)
	fs.writeFileSync(`./lib/${prefix}.json`, collection, 'utf8')
	const target = `./lib/${iconSet.prefix}`
	await exportJSONPackage(iconSet, {
		target,
		package: {
			name: `@rokkit/${iconSet.prefix}`,
			version: '1.0.0',
			bugs: 'https://github.com/jerrythomas/rokkit/issues',
			homepage: 'https://github.com/jerrythomas/rokkit'
		},
		cleanup: true
	})
}
