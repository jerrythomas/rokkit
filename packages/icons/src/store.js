import { writable } from 'svelte/store'
export const iconStore = writable({})
