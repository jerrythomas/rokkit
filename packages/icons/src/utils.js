export function toCapitalCase(text) {
	return text.charAt(0).toUpperCase() + text.slice(1).toLowerCase()
}

export function toPascalCase(text) {
	return text
		.split('-')
		.map((part) => toCapitalCase(part))
		.join('')
}

export function toHyphenCase(text) {
	return text
		.replace(/[A-Z]/g, (letter) => `-${letter.toLowerCase()}`)
		.replace(/^-/, '')
}

export function sortByParts(a, b) {
	const partsOfA = a.split('-')
	const partsOfB = b.split('-')

	let result =
		partsOfA[0] > partsOfB[0]
			? 1
			: partsOfA[0] < partsOfB[0]
			? -1
			: partsOfA.length - partsOfB.length
	return result != 0 ? result : a > b ? 1 : a < b ? -1 : 0
}
