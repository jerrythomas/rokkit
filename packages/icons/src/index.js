#!/usr/bin/env node
import { convert } from './convert.js'

const groups = ['base', 'components', 'auth', 'light', 'solid', 'twotone']

for (const group of groups) {
	convert(`./src/${group}`, group, group === 'auth')
}
