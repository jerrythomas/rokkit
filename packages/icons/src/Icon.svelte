<script>
	import { toPascalCase } from '@sparsh-ui/core'
	import { iconStore } from './store.js'

	export let size = '1.5em'
	export let name
	export let title = name
	export let fill = 'none'
	export let stroke = 'currentColor'
	export let isActive = false
	export let icon = $iconStore[toPascalCase(name)]
</script>

<svg
	xmlns="http://www.w3.org/2000/svg"
	class="{$$props.class ? $$props.class : ''} icon"
	class:isActive
	{fill}
	{stroke}
	height={size}
	width={size}
	viewBox="0 0 24 24"
	on:hover
	on:mouseover
	on:focus
	on:click
>
	<title>{title}</title>
	<svelte:component this={icon} />
</svg>

<style>
	svg {
		display: flex;
		margin: auto;
		flex-shrink: 0;
		flex-grow: 0;
	}
	svg:focus {
		outline: none;
	}
</style>
