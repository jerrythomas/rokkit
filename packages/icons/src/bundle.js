/**
 * This is an advanced example for creating icon bundles for Iconify SVG Framework.
 *
 * It creates a bundle from:
 * - All SVG files in a directory.
 * - Custom JSON files.
 * - Iconify icon sets.
 * - SVG framework.
 *
 * This example uses Iconify Tools to import and clean up icons.
 * For Iconify Tools documentation visit https://docs.iconify.design/tools/tools2/
 */
import { promises as fs } from 'fs'
import { dirname } from 'path'

// Installation: npm install --save-dev @iconify/tools @iconify/utils @iconify/json @iconify/iconify
import {
	importDirectory,
	cleanupSVG,
	parseColors,
	isEmptyColor,
	runSVGO
} from '@iconify/tools'
import { getIcons, stringToIcon, minifyIconSet } from '@iconify/utils'
//  import type { IconifyJSON, IconifyMetaData } from '@iconify/types';

/**
 * Script configuration
 */
//  interface BundleScriptCustomSVGConfig {
// 	 // Path to SVG files
// 	 dir: string;

// 	 // True if icons should be treated as monotone: colors replaced with currentColor
// 	 monotone: boolean;

// 	 // Icon set prefix
// 	 prefix: string;
//  }

//  interface BundleScriptCustomJSONConfig {
// 	 // Path to JSON file
// 	 filename: string;

// 	 // List of icons to import. If missing, all icons will be imported
// 	 icons?: string[];
//  }

//  interface BundleScriptConfig {
// 	 // Source file for Iconify SVG Framework:
// 	 // Use require.resolve('@iconify/iconify') for full version
// 	 // Use require.resolve('@iconify/iconify/dist/iconify.without-api.min') for version without API
// 	 svgFramework?: string;

// 	 // Custom SVG to import and bundle
// 	 svg?: BundleScriptCustomSVGConfig[];

// 	 // Icons to bundled from @iconify/json packages
// 	 iconifyIcons?: string[];

// 	 // List of JSON files to bundled
// 	 // Entry can be a string, pointing to filename or a BundleScriptCustomJSONConfig object (see type above)
// 	 // If entry is a string or object without 'icons' property, an entire JSON file will be bundled
// 	 json?: (string | BundleScriptCustomJSONConfig)[];
//  }

let sources //: BundleScriptConfig;
sources = {
	svgFramework: require.resolve(
		'@iconify/iconify/dist/iconify.without-api.min'
	),

	svg: [
		{
			dir: 'auth/monotone',
			monotone: true,
			prefix: 'auth'
		},
		{
			dir: 'auth/color',
			monotone: false,
			prefix: 'auth'
		},
		{
			dir: 'ui',
			monotone: false,
			prefix: 'sparsh'
		},
		{
			dir: 'preview',
			monotone: false,
			prefix: 'sparsh-ui'
		}
	]

	//  iconifyIcons: [
	// 	 'mdi:home',
	// 	 'mdi:account',
	// 	 'mdi:login',
	// 	 'mdi:logout',
	// 	 'octicon:book-24',
	// 	 'octicon:code-square-24',
	//  ],

	//  json: [
	// 	 // Custom JSON file
	// 	 'json/gg.json',
	// 	 // Iconify JSON file (@iconify/json is a package name, /json/ is directory where files are, then filename)
	// 	 require.resolve('@iconify/json/json/tabler.json'),
	// 	 // Custom file with only few icons
	// 	 {
	// 		 filename: require.resolve('@iconify/json/json/line-md.json'),
	// 		 icons: [
	// 			 'home-twotone-alt',
	// 			 'github',
	// 			 'document-list',
	// 			 'document-code',
	// 			 'image-twotone',
	// 		 ],
	// 	 },
	//  ],
}

// File to save bundle to
const target = 'assets/iconify-bundle.js'

/**
 * Do stuff!
 */
;(async function () {
	let bundle = ''

	// Create directory for output if missing
	const dir = dirname(target)
	try {
		await fs.mkdir(dir, {
			recursive: true
		})
	} catch (err) {
		//
	}

	/**
	 * Bundle SVG framework
	 */
	const isIconifyBundled = !!sources.svgFramework
	const wrapperFunction = isIconifyBundled ? 'Iconify.addCollection' : 'add'
	if (sources.svgFramework) {
		bundle += await fs.readFile(sources.svgFramework, 'utf8')
		console.log('Bundled SVG framework')

		// Try to copy .d.ts
		const tsSource = sources.svgFramework.replace('.js', '.d.ts')
		try {
			const tsContent = await fs.readFile(tsSource)
			await fs.writeFile(target.replace('.js', '.d.ts'), tsContent)
		} catch (err) {
			//
		}
	}

	/**
	 * Convert sources.iconifyIcons to sources.json
	 */
	//  if (sources.iconifyIcons) {
	// 	 const sourcesJSON = sources.json ? sources.json : (sources.json = []);

	// 	 // Sort icons by prefix
	// 	 const organizedList = organizeIconsList(sources.iconifyIcons);
	// 	 for (const prefix in organizedList) {
	// 		 const filename = require.resolve(`@iconify/json/json/${prefix}.json`);
	// 		 sourcesJSON.push({
	// 			 filename,
	// 			 icons: organizedList[prefix],
	// 		 });
	// 	 }
	//  }

	/**
	 * Bundle JSON files
	 */
	//  if (sources.json) {
	// 	 for (let i = 0; i < sources.json.length; i++) {
	// 		 const item = sources.json[i];

	// 		 // Load icon set
	// 		 const filename = typeof item === 'string' ? item : item.filename;
	// 		 let content = JSON.parse(
	// 			 await fs.readFile(filename, 'utf8')
	// 		 ) as IconifyJSON;

	// 		 // Filter icons
	// 		 if (typeof item !== 'string' && item.icons?.length) {
	// 			 const filteredContent = getIcons(content, item.icons);
	// 			 if (!filteredContent) {
	// 				 throw new Error(`Cannot find required icons in ${filename}`);
	// 			 }
	// 			 content = filteredContent;
	// 		 }

	// 		 // Remove metadata and add to bundle
	// 		 removeMetaData(content);
	// 		 minifyIconSet(content);
	// 		 bundle += wrapperFunction + '(' + JSON.stringify(content) + ');\n';
	// 		 console.log(`Bundled icons from ${filename}`);
	// 	 }
	//  }

	/**
	 * Custom SVG
	 */
	if (sources.svg) {
		for (let i = 0; i < sources.svg.length; i++) {
			const source = sources.svg[i]

			// Import icons
			const iconSet = await importDirectory(source.dir, {
				prefix: source.prefix
			})

			// Validate, clean up, fix palette and optimise
			await iconSet.forEach(async (name, type) => {
				if (type !== 'icon') {
					return
				}

				// Get SVG instance for parsing
				const svg = iconSet.toSVG(name)
				if (!svg) {
					// Invalid icon
					iconSet.remove(name)
					return
				}

				// Clean up and optimise icons
				try {
					// Clean up icon code
					await cleanupSVG(svg)

					if (source.monotone) {
						// Replace color with currentColor, add if missing
						// If icon is not monotone, remove this code
						await parseColors(svg, {
							defaultColor: 'currentColor',
							callback: (attr, colorStr, color) => {
								return !color || isEmptyColor(color) ? colorStr : 'currentColor'
							}
						})
					}

					// Optimise
					await runSVGO(svg)
				} catch (err) {
					// Invalid icon
					console.error(`Error parsing ${name} from ${source.dir}:`, err)
					iconSet.remove(name)
					return
				}

				// Update icon from SVG instance
				iconSet.fromSVG(name, svg)
			})
			console.log(`Bundled ${iconSet.count()} icons from ${source.dir}`)

			// Export to JSON
			const content = iconSet.export()
			bundle += wrapperFunction + '(' + JSON.stringify(content) + ');\n'
		}
	}

	/**
	 * Add wrapper function if SVG framework is not in bundle
	 */
	if (!isIconifyBundled) {
		// Wrap in custom code that checks for Iconify.addCollection and IconifyPreload
		bundle = `(function() {
	function add(data) {
		try {
			if (typeof self.Iconify === 'object' && self.Iconify.addCollection) {
				self.Iconify.addCollection(data);
				return;
			}
			if (typeof self.IconifyPreload === 'undefined') {
				self.IconifyPreload = [];
			}
			self.IconifyPreload.push(data);
		} catch (err) {
		}
	}
	${bundle}
	})();\n`
	}

	// Save to file
	await fs.writeFile(target, bundle, 'utf8')

	console.log(`Saved ${target} (${bundle.length} bytes)`)
})().catch((err) => {
	console.error(err)
})

/**
 * Remove metadata from icon set
 */
//  function removeMetaData(iconSet: IconifyJSON) {
// 	 const props: (keyof IconifyMetaData)[] = [
// 		 'info',
// 		 'chars',
// 		 'categories',
// 		 'themes',
// 		 'prefixes',
// 		 'suffixes',
// 	 ];
// 	 props.forEach((prop) => {
// 		 delete iconSet[prop];
// 	 });
//  }

/**
 * Sort icon names by prefix
 */
//  function organizeIconsList(icons: string[]): Record<string, string[]> {
// 	 const sorted: Record<string, string[]> = Object.create(null);
// 	 icons.forEach((icon) => {
// 		 const item = stringToIcon(icon);
// 		 if (!item) {
// 			 return;
// 		 }

// 		 const prefix = item.prefix;
// 		 const prefixList = sorted[prefix] ? sorted[prefix] : (sorted[prefix] = []);

// 		 const name = item.name;
// 		 if (prefixList.indexOf(name) === -1) {
// 			 prefixList.push(name);
// 		 }
// 	 });

// 	 return sorted;
//  }
