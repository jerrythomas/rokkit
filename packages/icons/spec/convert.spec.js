import { describe, it, expect, beforeAll, afterAll } from 'vitest'
import { convert } from '../src/convert'
import fs from 'fs'
import { omit } from 'ramda'
import { rimraf } from 'rimraf'
// Helper function to read JSON files
const readJSON = (filePath) => JSON.parse(fs.readFileSync(filePath, 'utf8'))

describe('convert', () => {
	const originalCwd = process.cwd()
	const baseDir = 'packages/icons/spec/fixtures'

	beforeAll(() => {
		process.chdir(baseDir)
		fs.mkdirSync('lib', { recursive: true })
	})

	afterAll(() => {
		rimraf.sync('./lib')
		// cleanOutputDir('./lib')
		process.chdir(originalCwd)
	})

	it('should import, process, and export icons correctly for b/w icons', async () => {
		// const inputDir = path.join(baseDir, 'icons', 'app')
		const prefix = 'app'
		await convert('./src/app', prefix)
		const files = fs.readdirSync('./lib/app')
		expect(files).toMatchSnapshot()
		const icons = readJSON('./lib/app.json')
		expect(omit(['lastModified'], icons)).toMatchSnapshot()
	})

	it('should import, process, and export icons correctly for colored icons', async () => {
		// const inputDir = path.join(baseDir, 'icons', 'web')
		const prefix = 'web'
		// const expectedOutput = readJSON(path.join(__dirname, 'fixtures', 'expected', 'web.json'))

		await convert('./src/web', prefix, true)
		const files = fs.readdirSync('./lib/web')
		expect(files).toMatchSnapshot()
		const icons = readJSON('./lib/web.json')
		expect(omit(['lastModified'], icons)).toMatchSnapshot()
	})

	it('should exclude invalid files', async () => {
		const prefix = 'err1'
		// eslint-disable-next-line no-console
		console.error = vi.fn()
		await convert('./src/err1', prefix)

		// eslint-disable-next-line no-console
		expect(console.error).toHaveBeenCalledWith('Error parsing not-icon: Invalid attribute name')
		// console.error = originalConsoleError
		await convert('./src/err1', prefix)
		const files = fs.readdirSync('./lib/err1')
		expect(files).toMatchSnapshot()
		const icons = readJSON('./lib/err1.json')
		expect(omit(['lastModified'], icons)).toMatchSnapshot()
		vi.restoreAllMocks()
	})
})
