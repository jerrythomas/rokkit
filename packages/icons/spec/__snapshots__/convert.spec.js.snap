// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convert > should exclude invalid files 1`] = `
[
  "chars.json",
  "icons.json",
  "index.d.ts",
  "index.js",
  "index.mjs",
  "info.json",
  "metadata.json",
  "package.json",
]
`;

exports[`convert > should exclude invalid files 2`] = `
{
  "icons": {},
  "prefix": "err1",
}
`;

exports[`convert > should import, process, and export icons correctly for b/w icons 1`] = `
[
  "chars.json",
  "icons.json",
  "index.d.ts",
  "index.js",
  "index.mjs",
  "info.json",
  "metadata.json",
  "package.json",
]
`;

exports[`convert > should import, process, and export icons correctly for b/w icons 2`] = `
{
  "icons": {
    "alert-bell": {
      "body": "<g fill="none" stroke="currentColor" stroke-width="1.5"><path d="M18.75 9.71v-.705C18.75 5.136 15.726 2 12 2S5.25 5.136 5.25 9.005v.705a4.4 4.4 0 01-.692 2.375L3.45 13.81c-1.011 1.575-.239 3.716 1.52 4.214a25.8 25.8 0 0014.06 0c1.759-.498 2.531-2.639 1.52-4.213l-1.108-1.725a4.4 4.4 0 01-.693-2.375Z"/><path stroke-linecap="round" d="M7.5 19c.655 1.748 2.422 3 4.5 3s3.845-1.252 4.5-3" opacity=".5"/></g>",
      "height": 24,
      "width": 24,
    },
  },
  "prefix": "app",
}
`;

exports[`convert > should import, process, and export icons correctly for colored icons 1`] = `
[
  "chars.json",
  "icons.json",
  "index.d.ts",
  "index.js",
  "index.mjs",
  "info.json",
  "metadata.json",
  "package.json",
]
`;

exports[`convert > should import, process, and export icons correctly for colored icons 2`] = `
{
  "icons": {
    "authy": {
      "body": "<g fill="none"><path fill="#F22E46" d="M24 12c0 6.628-5.372 12-12 12S0 18.628 0 12 5.372 0 12 0s12 5.372 12 12"/><path fill="#fff" d="m10.971 9.458 2.877 2.877a1.065 1.065 0 001.506-1.506l-2.877-2.877a5.38 5.38 0 00-7.546-.053l-.026.027-.027.026-.051.056a5.38 5.38 0 00.104 7.49l2.878 2.877a1.065 1.065 0 001.506-1.506l-2.877-2.877a3.246 3.246 0 01-.026-4.56 3.246 3.246 0 014.559.027"/><path fill="#F5B1B2" d="M14.685 5.627a1.063 1.063 0 000 1.505l2.877 2.878a3.247 3.247 0 01.025 4.56 3.246 3.246 0 01-4.558-.027l-2.878-2.877a1.065 1.065 0 10-1.506 1.506l2.877 2.877a5.38 5.38 0 007.49.105l.056-.052q.014-.013.026-.027l.027-.027q.027-.026.05-.055a5.38 5.38 0 00-.102-7.49L16.19 5.627a1.064 1.064 0 00-1.506 0"/></g>",
      "height": 24,
      "width": 24,
    },
  },
  "prefix": "web",
}
`;
